src/constraint/4C_constraint_springdashpot.cpp

```diff
@@ -10,21 +10,228 @@
 #include "4C_truss3.hpp"
+#include "4C_utils_exceptions.hpp"
 #include "4C_utils_function.hpp"
 #include "4C_utils_function_of_time.hpp"
 
-#include <iostream>
+#include <algorithm>
+#include <concepts>
+#include <tuple>
 #include <utility>
 
 FOUR_C_NAMESPACE_OPEN
 
+namespace
+{
+
+  double scale_by_optional_function(double baseline_value, int num_func, double total_time)
+  {
+    return num_func ? Global::Problem::instance()
+                              ->function_by_id<Core::Utils::FunctionOfTime>(num_func)
+                              .evaluate(total_time) *
+                          baseline_value
+                    : baseline_value;
+  }
+
+  template <typename T>
+  concept SpringDashpotEvaluatable = requires(T t, double delta_displacement, double velocity) {
+    { t(delta_displacement, velocity) } -> std::convertible_to<std::tuple<double, double, double>>;
+  };
+
+  template <SpringDashpotEvaluatable SpringDashpot, Core::FE::CellType celltype, unsigned int dim>
+  void add_robin_spring_dashpot_contribution_xyz(const Core::Elements::Element& surface_element,
+      const Core::Elements::ElementNodes<celltype, dim> element_nodes,
+      const Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, dim>& displacements,
+      const Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, dim>& displacement_offset,
+      const std::vector<double>& constant_offset,
+      const Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, dim>& velocity,
+      const std::vector<SpringDashpot>& spring_dashpot_evaluator, const double time_factor,
+      const Core::FE::GaussIntegration& gauss_integration, const std::vector<bool>& onoff,
+      std::optional<Core::LinAlg::Matrix<dim * Core::FE::num_nodes<celltype>, 1>>& residual_vector,
+      std::optional<Core::LinAlg::Matrix<dim * Core::FE::num_nodes<celltype>,
+          dim * Core::FE::num_nodes<celltype>>>& stiffness_matrix)
+  {
+    Core::Elements::for_each_gauss_point<celltype>(element_nodes, gauss_integration,
+        [&](const Core::LinAlg::Matrix<Core::FE::dim<celltype>, 1>& xi,
+            const Core::Elements::ShapeFunctionsAndDerivatives<celltype>& shape_functions,
+            const Core::Elements::JacobianMapping<celltype, dim>& jacobian_mapping,
+            double integration_factor, int gp)
+        {
+          Core::LinAlg::Matrix<dim, 1> displacement_at_gp;
+          displacement_at_gp.multiply_tn(displacements, shape_functions.values);
+
+          Core::LinAlg::Matrix<dim, 1> offset_at_gp;
+          offset_at_gp.multiply_tn(displacement_offset, shape_functions.values);
+
+          Core::LinAlg::Matrix<dim, 1> velocity_at_gp;
+          velocity_at_gp.multiply_tn(velocity, shape_functions.values);
+
+          for (unsigned axis_direction = 0; axis_direction < dim; ++axis_direction)
+          {
+            if (!onoff[axis_direction]) continue;
+
+            FOUR_C_ASSERT(spring_dashpot_evaluator.size() > axis_direction,
```

Comment:
```suggestion
            FOUR_C_ASSERT(axis_direction < spring_dashpot_evaluator.size(),
```

reads a lot better, IMO.